docs/chymyst07.md

```diff
@@ -0,0 +1,731 @@
+The only way to prevent a reaction from starting is to omit an input molecule.
+Therefore, we need to introduce another molecule (say, `beganOnce`) as input into that reaction.
+The reaction will consume that molecule and never emit it again.
+
+```scala
+val beginCritical = b[Unit, M[Unit]]
+val access = m[Unit]
+
+site (
+  go { case beginCritical(_, reply) + access(_) =>
+    val endCritical = m[Unit] // Declare a new emitter.
+    val beganOnce = m[Unit]
+    site (
+      go { case endCritical(_) + beganOnce(_) => access() }
+    )
+    beganOnce() // Emit only one copy.
+    reply(endCritical) // beginCritical() returns the new emitter.
+  }
+)
+access() // Emit only one copy.
+
+// Example usage:
+val endCritical = beginCritical()
+
+???... // The code of the critical section.
+
+endCritical() // End of the critical section.
+
+endCritical() // This has no effect because `beganOnce()` is not available any more.
+
+```
+
+As before, we can easily modify this code to support multiple (but limited) concurrent entry into critical sections or token-based access. 
+
+### Refactoring into a library
+
+We would like to be able to create new, unique `beginCritical()` molecules on demand, so that we could have multiple distinct critical sections in our code.
+
+For instance, we could declare two critical sections and use them in three reactions that could run concurrently like this: 
+
+```scala
+val beginCritical1 = newCriticalSectionMarker()
+val beginCritical2 = newCriticalSectionMarker()
+
+```
+
+| Reaction 1 | |  Reaction 2 | | Reaction 3 |
+|---|---|---|---|---|
+| `beginCritical1()` | | ... | | `beginCritical2()` |
+| `beginCritical2()` | | ... | | ... |
+| (blocked by 3) | | `beginCritical1()` | | ... |
+| (starts running) | | (blocked by 1) | | `endCritical2()` |
+| ... | |  (still blocked)| | `beginCritical2()` |
+| `endCritical1()` | | (starts running) | | (blocked by 1) |
+| `endCritical2()` | | ... | | (starts running) |
+| ... | | ... | | ... |
+| ... | | `endCritical1()` | | `endCritical2()` |
+
+The result should be that we can create any number of critical sections that work independently of each other.
+
+In order to package the implementation of the critical section into a library function `newCriticalSectionMarker`, we simply declare the chemistry in the local scope of that function and return the molecule emitter:
+
+```scala
+def newCriticalSectionMarker(): B[Unit, M[Unit]] = {
+
+  val beginCritical = b[Unit, M[Unit]]
+  val access = m[Unit]
+
+  site (
+    go { case beginCritical(_, reply) + access(_) =>
+      val endCritical = m[Unit] // Declare a new emitter.
+      val beganOnce = m[Unit]
+      site (
+        go { case endCritical(_) + beganOnce(_) => access() }
+      )
+      beganOnce() // Emit only one copy.
+      reply(endCritical) // beginCritical() returns the new emitter.
+    }  
+  )
+  access() // Emit only one copy.
+
+  beginCritical // Return the new emitter.  
+}
+
+// Example usage:
```

Comment:
It is not easy to come up with an example. What is an example of a critical section that can occur in different places in the code?
I'm not sure why you say that "we consume beginCritical1 and assign that to endCritical1".
Actually, we _emit_ a `beginCritical1()` molecule and, since this is a blocking molecule, its emitter will block and return a value, which is the `endCritical1` emitter. Later we will emit `endCritical1()`.